### PR TITLE
Implement Bulk Version Two Protocol

### PR DESCRIPTION
**Pull request by Dala0**
## What does this change?

Implements the use of the bulk version two protocol.

## How to test

Build the executable with Xcode and try to download a bulk from the server. The bulk should download successfully.

## How can we measure success?

Bulks with many files should now download successfully.

### CODE REVIEW

File: ArchiveHunterDownloadManager/AppDelegate.m [hunk 1]
```diff
@@ -429,4 +446,25 @@ - (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication *)sende
     return NSTerminateNow;
 }
 
+- (void) getEntities:(NSManagedObject *)bulk {
+    NSString *token = [self getReterievalToken:bulk];
+    NSString *serverType = [self getServerType:bulk];
+    NSString *serverHost = [self hostNameForServerSource:serverType];
+    NSURL *apiURL = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/api/bulkv2/%@/summarystream", serverHost, token]];
```
Comment (fredex42) [inline ArchiveHunterDownloadManager/AppDelegate.m:458]: From the function definitions above, `*token` and `*serverType` are both nullable - so  before using them here you should check if either is null and return early from `getEntities`, logging out a message with NSLog

File: ArchiveHunterDownloadManager/AppDelegate.m [hunk 2]
```diff
@@ -177,28 +177,17 @@ - (void) initialiseDownloadFromUrl:(NSString *)token forServerType:(NSString *)s
...
                 }
                 
-                NSDictionary *entryDict = [data objectForKey:@"entries"];
-                NSUInteger keyCount = [entryDict count];
-                
-                if(keyCount==0) {
+                NSNumber *totalFileSize = [bulk valueForKey:@"totalSize"];
+                long long fileSizeAsLongLong = [totalFileSize longLongValue];
+
+                if(fileSizeAsLongLong==0) {
```
Comment (fredex42) [inline ArchiveHunterDownloadManager/AppDelegate.m]: Would be simpler to write this as:

```objective-c
if( [ [bulk valueForKey:@"totalSize"] isEqualToNumber:[NSNumber numberWithInt:0] ]) {
```

File: ArchiveHunterDownloadManager/AppDelegate.m [hunk 3]
```diff
@@ -177,28 +177,17 @@ - (void) initialiseDownloadFromUrl:(NSString *)token forServerType:(NSString *)s
...
-                    [self createNewEntry:entrySynop parent:bulk];
-                    NSLog(@"entry synopsis is %@", entrySynop);
-                    NSNumber *fileSize = [entrySynop valueForKey:@"fileSize"];
-                    //NSLog(@"File Size: %@", fileSize);
-                    //NSLog(@"File Size Long Long: %lld", [fileSize longLongValue]);
-                    totalSize = [fileSize longLongValue] + totalSize;
-                }
-                
-                [bulk setValue:[NSNumber numberWithLongLong:totalSize] forKey:@"totalSize"];
+                [self getEntities:bulk];
```
Comment (fredex42) [inline ArchiveHunterDownloadManager/AppDelegate.m:180]: 👍 Good to put this in its own function 😃

**Review by fredex42 — CHANGES_REQUESTED** [review]
Just one minor change, good work :)

File: ArchiveHunterDownloadManager/AppDelegate.m (hunk 1 above)
Comment (Dala0) [reply ArchiveHunterDownloadManager/AppDelegate.m:458]: Done.

File: ArchiveHunterDownloadManager/AppDelegate.m (hunk 2 above)
Comment (Dala0) [reply ArchiveHunterDownloadManager/AppDelegate.m]: Done.

**Review by fredex42 — APPROVED** [review]
:+1:

**Comment by fredex42** [comment]
now we just have to get a build....

**Pull request merged**